### PR TITLE
TEST-#5350: port DropDuplicates and LevelAlign benchmarks from pandas github 

### LINKED ISSUE #5350: [ASV] port align and DropDuplicates benchmarks from pandas github
Have to add ASV benchmarks related which does the following actions

1)Dataframe.align and Dataframe.reindex

2)drop_duplicates method for Dataframe and Series. 
        - With separate benchmarks for inplace argument true and false 
        - With different types of data such as  string,integer ,nan

### PR DESCRIPTION
**Pull request by arunjose696**
## What do these changes do?

Adding 2 benchmark classes from pandas benchmark
Source PR https://github.com/modin-project/modin/pull/5047

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5350 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: asv_bench/benchmarks/benchmarks.py [hunk 1]
```diff
@@ -992,4 +993,96 @@ def time_float_32(self, shape, method):
...
+        execute(self.df)
+        execute(self.df_level)
+
+    def time_align_level(self, shape):
+        left, right = self.df.align(self.df_level, level=1, copy=False)
+        execute(left)
+        execute(right)
+
+    def time_reindex_level(self, shape):
+        execute(self.df_level.reindex(self.index, level=1))
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1026]: This looks like out of scope of `TimeLevelAlign` class.

File: asv_bench/benchmarks/benchmarks.py [hunk 2]
```diff
@@ -992,4 +993,96 @@ def time_float_32(self, shape, method):
         execute(self.df_ts_float32.fillna(method=method))
 
 
+class TimeLevelAlign:
+    params = [get_benchmark_shapes("TimeLevelAlign")]
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:997]: It's probably not a good idea to use the forms for the unary operation that supposed to be used for the binary operations.

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/benchmarks.py (hunk 2 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py:997]: As Align  aligns two objects. I was thinking this would be a binary operation.

Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py:997]: oh, you are right! ok

File: asv_bench/benchmarks/benchmarks.py (hunk 1 above)
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1026]:
```suggestion
        self.df1 = IMPL.DataFrame(
            np.random.randn(len(self.index), col1), index=self.index
        )
        self.df2 = IMPL.DataFrame(np.random.randn(row2, col2))
        execute(self.df1), execute(self.df2)

    def time_align_level(self, shapes):
        left, right = self.df1.align(self.df2, level=1, copy=False)
        execute(left), execute(right)
```

File: asv_bench/benchmarks/benchmarks.py [hunk 3]
```diff
@@ -992,4 +993,96 @@ def time_float_32(self, shape, method):
...
+    param_names = ["shape"]
+
+    def setup(self, shape):
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1000]:
```suggestion
    param_names = ["shapes"]

    def setup(self, shapes):
```

File: asv_bench/benchmarks/benchmarks.py [hunk 4]
```diff
@@ -992,4 +993,96 @@ def time_float_32(self, shape, method):
...
+    def time_align_level(self, shape):
+        left, right = self.df.align(self.df_level, level=1, copy=False)
+        execute(left)
+        execute(right)
+
+    def time_reindex_level(self, shape):
+        execute(self.df_level.reindex(self.index, level=1))
+
+
+class TimeDropDuplicates:
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1029]: Let's split it into two classes: `TimeDropDuplicatesDataFrame`, `TimeDropDuplicatesSeries`.

File: asv_bench/benchmarks/benchmarks.py (hunk 1 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py:1026]: In pandas ASV benchmarks they have added this function in LevelAlign class. What I figured was df.align method and df.reindex by providing index of other dataframe produces the same result. So these were two types of aligning a dataframe. 
Should I remove this method from benchmark or keep it to make it similar to pandas benchmarks?

Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py:1026]: Ok, you can leave it with a comment. Something like: "`reindex` returns the same result as `align`. Approximately the same performance is expected."

Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: done

File: asv_bench/benchmarks/benchmarks.py (hunk 3 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: done

File: asv_bench/benchmarks/benchmarks.py (hunk 4 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: done

File: asv_bench/benchmarks/benchmarks.py [hunk 5]
```diff
@@ -992,4 +993,106 @@ def time_float_32(self, shape, method):
...
+        execute(self.df_int)
+
+
+class TimeDropDuplicatesSeries:
+
+    params = [get_benchmark_shapes("TimeDropDuplicatesSeries")]
+    param_names = ["shape"]
+
+    def setup(self, shape):
+        N = shape[0]
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1078]:
```suggestion
        rows = shape[0]
```

File: asv_bench/benchmarks/benchmarks.py [hunk 6]
```diff
@@ -992,4 +993,106 @@ def time_float_32(self, shape, method):
...
+        execute(self.s.drop_duplicates())
+
+    def time_series_drop_dups_int_inplace(self, shape):
+        self.s.drop_duplicates(inplace=True)
+        execute(self.s)
+
+    def time_series_drop_dups_string(self, shape):
+        execute(self.s_str.drop_duplicates())
+
+    def time_series_drop_dups_string_inplace(self, shape):
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1093]:
```suggestion
    def time_drop_dups_int(self, shape):
        execute(self.s.drop_duplicates())

    def time_drop_dups_int_inplace(self, shape):
        self.s.drop_duplicates(inplace=True)
        execute(self.s)

    def time_drop_dups_string(self, shape):
        execute(self.s_str.drop_duplicates())

    def time_drop_dups_string_inplace(self, shape):
```

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/benchmarks.py [hunk 7]
```diff
@@ -992,4 +993,106 @@ def time_float_32(self, shape, method):
...
+        self.s = IMPL.Series(np.random.randint(0, N // 10, size=N))
+        self.s_str = IMPL.Series(np.tile(tm.makeStringIndex(N // 10).values, 10))
+        execute(self.s), execute(self.s_str)
+
+    def time_series_drop_dups_int(self, shape):
+        execute(self.s.drop_duplicates())
+
+    def time_series_drop_dups_int_inplace(self, shape):
+        self.s.drop_duplicates(inplace=True)
+        execute(self.s)
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1088]: Modin does not have a specific implementation for different types for `drop_duplicates`. I propose to leave one type here. For example, the string type.

Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py:1088]: Same for `TimeDropDuplicatesDataframe` class.

Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: Removed all drop_duplicates function except for strings

File: asv_bench/benchmarks/benchmarks.py (hunk 5 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: Done

File: asv_bench/benchmarks/benchmarks.py [hunk 8]
```diff
@@ -992,4 +993,106 @@ def time_float_32(self, shape, method):
...
+
+
+class TimeDropDuplicatesDataframe:
+
+    params = [get_benchmark_shapes("TimeDropDuplicatesDataframe")]
+    param_names = ["shape"]
+
+    def setup(self, shape):
+        N = shape[0]
+        K = shape[1]
```
[the remark this answers is not in the document]
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: As N and K are not rows and cols but K is the number of repetations, so now K has been replaced with a scalar value 10 so N*K will be the number of rows obtained from shape parameter

File: asv_bench/benchmarks/benchmarks.py [hunk 9]
```diff
@@ -992,4 +993,78 @@ def time_float_32(self, shape, method):
...
+
+    params = [get_benchmark_shapes("TimeDropDuplicatesDataframe")]
+    param_names = ["shape"]
+
+    def setup(self, shape):
+        N = shape[0] // 10
+        K = 10
+        key1 = tm.makeStringIndex(N).values.repeat(K)
+        key2 = tm.makeStringIndex(N).values.repeat(K)
+        self.df = IMPL.DataFrame(
```
[the remark this answers is not in the document]
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: self.df now has dimensions as in shape

File: asv_bench/benchmarks/benchmarks.py [hunk 10]
```diff
@@ -992,4 +993,79 @@ def time_float_32(self, shape, method):
...
+        K = 10
+        self.df = IMPL.DataFrame()
+        # dataframe would  have cols-1 keys(strings) and one value(int) column
+        for col in range(cols - 1):
+            self.df["key" + str(col + 1)] = tm.makeStringIndex(N).values.repeat(K)
+        self.df["value"] = np.random.randn(N * K)
+        execute(self.df)
+
+    def time_drop_dups(self, shape):
+        execute(self.df.drop_duplicates(["key1", "key2"]))
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1046]: Let's perform the operation on all columns:
```suggestion
        execute(self.df.drop_duplicates())
```

File: asv_bench/benchmarks/benchmarks.py [hunk 11]
```diff
@@ -992,4 +993,79 @@ def time_float_32(self, shape, method):
...
+
+    def time_drop_dups_inplace(self, shape):
+        self.df.drop_duplicates(["key1", "key2"], inplace=True)
```
Comment (anmyachev) [inline asv_bench/benchmarks/benchmarks.py:1049]: Same
```suggestion
        self.df.drop_duplicates(inplace=True)
```

Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: done

File: asv_bench/benchmarks/benchmarks.py (hunk 10 above)
Comment (arunjose696) [reply asv_bench/benchmarks/benchmarks.py]: done

File: asv_bench/benchmarks/benchmarks.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (anmyachev) [reply asv_bench/benchmarks/benchmarks.py]: thanks @arunjose696

**Review by anmyachev — APPROVED** [review]
LGTM!
